src/devices/cpu/z180/z180asci.cpp

```diff
@@ -0,0 +1,672 @@
+	{
+		// External clock
+		m_bgr_divisor = 0;
+	}
+	else
+	{
+		if (m_asci_ext & Z180_ASEXT_BRG0_MODE)
+		{
+			// Extended boud rate generator mode
+			m_bgr_divisor = m_asci_tc.w + 2;
+		}
+		else
+		{
+			// Regular bitrate generator mode
+			m_bgr_divisor = 1 << (m_asci_cntlb & Z180_CNTLB_SS);
+			m_bgr_divisor *= ((m_asci_cntlb & Z180_CNTLB_CTS_PS) ? 30 : 10); // Prescale
+		}
+	}
+
+	if (m_bgr_divisor)
+	{
+		LOG("Z180 ASCI%d set bitrate %d\n", m_id, uint32_t(clock() / m_bgr_divisor / m_divisor));
+		m_brg->set_clock(DERIVED_CLOCK(1,m_bgr_divisor));
+	}
+	else
+	{
+		LOG("Z180 ASCI%d set bitrate 0, using external\n", m_id);
+		m_brg->set_clock(0);
+	}
+}
+
+uint8_t z180asci_channel_base::cntla_r()
+{
+	LOG("Z180 CNTLA%d rd $%02x\n", m_id, m_asci_cntla);
+	return m_asci_cntla;
+}
+
+uint8_t z180asci_channel_base::cntlb_r()
+{
+	uint8_t data = (m_asci_cntlb & 0x0d) | (m_cts << 5);
+	LOG("Z180 CNTLB%d rd $%02x\n", m_id, data);
+	return data;
+}
+
+uint8_t z180asci_channel_base::stat_r()
+{
+	LOG("Z180 STAT%d  rd $%02x\n", m_id, m_asci_stat);
+	return m_asci_stat;
+}
+
+uint8_t z180asci_channel_base::tdr_r()
+{
+	LOG("Z180 TDR%d   rd $%02x\n", m_id, m_asci_tdr);
+	return m_asci_tdr;
+}
+
+uint8_t z180asci_channel_base::rdr_r()
+{
+	LOG("Z180 RDR%d   rd $%02x\n", m_id, m_asci_rdr);
+	if (!machine().side_effects_disabled())
+	{
+		if (m_fifo_rd != m_fifo_wr)
+		{
+			m_asci_rdr = m_data_fifo[m_fifo_rd];
+			m_asci_stat &= ~(Z180_STAT_OVRN | Z180_STAT_PE | Z180_STAT_FE);
+			m_asci_stat |= m_error_fifo[m_fifo_rd];
+			if (m_asci_stat & (Z180_STAT_OVRN | Z180_STAT_PE | Z180_STAT_FE))
+				m_irq = 1;
+			m_fifo_rd = (m_fifo_rd + 1) & 3;
+			if (m_fifo_rd == m_fifo_wr) // empty
+				m_asci_stat &= ~Z180_STAT_RDRF;
+		}
+	}
+	return m_asci_rdr;
+}
+
+uint8_t z180asci_channel_base::asext_r()
+{
+	LOG("Z180 ASEXT%d rd $%02x\n", m_id, m_asci_ext);
+	return m_asci_ext;
+}
+
+uint8_t z180asci_channel_base::astcl_r()
+{
+	LOG("Z180 ASTC%dL rd $%02x ($%04x)\n", m_id, m_asci_tc.b.l, m_asci_tc.w);
+	return m_asci_tc.b.l;
+}
+
+uint8_t z180asci_channel_base::astch_r()
+{
+	LOG("Z180 ASTC%dH rd $%02x ($%04x)\n", m_id, m_asci_tc.b.h, m_asci_tc.w);
+	return m_asci_tc.b.h;
+}
+
+void z180asci_channel_base::cntla_w(uint8_t data)
+{
+	LOG("Z180 CNTLA%d wr $%02x\n", m_id, data);
+	m_asci_cntla = data & ~(Z180_CNTLA_MPBR_EFR | Z180_CNTLA_RTS0);
+	output_rts(BIT(data,4)); // Z180_CNTLA_RTS0
+	if (data & Z180_CNTLA_MPBR_EFR) // Error Flag Reset
+	{
+		m_asci_stat &= ~(Z180_STAT_OVRN | Z180_STAT_PE | Z180_STAT_FE);
+		m_asci_ext &= ~(Z180_ASEXT_BRK_DET);
+	}
+}
+
+void z180asci_channel_base::cntlb_w(uint8_t data)
+{
+	LOG("Z180 CNTLB%d wr $%02x\n", m_id, data);
+	m_asci_cntlb = data;
+	device_clock_changed();
+}
+
+void z180asci_channel_base::tdr_w(uint8_t data)
+{
+	LOG("Z180 TDR%d   wr $%02x\n", m_id, data);
+	m_asci_tdr = data;
+	m_asci_stat &= ~Z180_STAT_TDRE;
+	m_tx_bits = 0;
+	m_tx_state = STATE_START;
+	m_tsr = m_asci_tdr;
+}
```

Comment:
This won’t work right – if the framing or MP bit are changed during transmission, it should not affect the byte being transmitted.

The way these things work in real life is they have a shift register one bit longer than the maximum length of a fully framed byte.  When loading the transmit shift register, it loads a bit pattern with a zero bit marking the end framed byte.  As the byte is shifted out, one bits are shifted in from the most significant bit.  It knows it needs to stop shifting bits out by detecting when the shift register contains a single zero bit in the least significant position.  If you emulate this properly, you can avoid the complex switch statement when shifting bits out, and you don’t need a bit counter at all.

Also, the transmit shift register and transmit data register are separate.  You can write a byte to the transmit data register without losing the contents of the transmit shift register – it will be transferred to the transmit shift register when the current content has finished shifting out.